tachyon/crypto/hashes/sponge/poseidon2/poseidon2_plonky3_internal_matrix.h

```diff
@@ -7,50 +7,74 @@
+#include "tachyon/math/finite_fields/finite_field_traits.h"
 #include "tachyon/math/matrix/matrix_types.h"
 
 namespace tachyon::crypto {
 
-template <typename PrimeField>
+template <typename F>
 class Poseidon2Plonky3InternalMatrix {
  public:
-  static void Apply(math::Vector<PrimeField>& v,
-                    const math::Vector<uint8_t>& shifts) {
+  template <typename F2 = F, std::enable_if_t<math::FiniteFieldTraits<
+                                 F2>::kIsPackedPrimeField>* = nullptr>
+  static void Apply(math::Vector<F>& v,
+                    const math::Vector<F>& diagonal_minus_one) {
+    using PrimeField =
+        std::conditional_t<math::FiniteFieldTraits<F>::kIsPackedPrimeField,
+                           typename math::FiniteFieldTraits<F>::PrimeField, F>;
+
+    Poseidon2HorizenInternalMatrix<F>::Apply(v, diagonal_minus_one);
+    if constexpr (PrimeField::Config::kUseMontgomery) {
+      static_assert(PrimeField::Config::kModulusBits <= 32);
+      for (F& f : v) {
+        f *= F::Broadcast(PrimeField::FromMontgomery(1));
+      }
+    }
+  }
+
+  template <
+      typename F2 = F,
+      std::enable_if_t<math::FiniteFieldTraits<F2>::kIsPrimeField>* = nullptr>
+  static void Apply(math::Vector<F>& v, const math::Vector<uint8_t>& shifts) {
     // |partial_sum| =       v₁ + v₂ + ... + vₙ₋₂ + vₙ₋₁
     // |full_sum|    =  v₀ + v₁ + v₂ + ... + vₙ₋₂ + vₙ₋₁
     //       s₀      = -v₀ + v₁ + v₂ + ... + vₙ₋₂ + vₙ₋₁
     //       sᵢ      = |full_sum| + (vᵢ << shiftsᵢ₋₁)
-    static_assert(PrimeField::Config::kModulusBits <= 32);
+    static_assert(F::Config::kModulusBits <= 32);
     uint64_t partial_sum = std::accumulate(
         v.begin() + 1, v.end(), uint64_t{0},
-        [](uint64_t acc, PrimeField value) { return acc += value.value(); });
+        [](uint64_t acc, F value) { return acc += value.value(); });
     uint64_t full_sum = partial_sum + v[0].value();
     uint64_t s0 = partial_sum + (-v[0]).value();
 
-    if constexpr (PrimeField::Config::kUseMontgomery) {
-      v[0] = PrimeField::FromMontgomery(PrimeField::Config::FromMontgomery(s0));
+    if constexpr (F::Config::kUseMontgomery) {
+      v[0] = F::FromMontgomery(F::Config::FromMontgomery(s0));
     } else {
       v[0] = FromU62(s0);
     }
     for (Eigen::Index i = 1; i < v.size(); ++i) {
       uint64_t si = full_sum + (uint64_t{v[i].value()} << shifts[i - 1]);
-      if constexpr (PrimeField::Config::kUseMontgomery) {
-        v[i] =
-            PrimeField::FromMontgomery(PrimeField::Config::FromMontgomery(si));
+      if constexpr (F::Config::kUseMontgomery) {
+        v[i] = F::FromMontgomery(F::Config::FromMontgomery(si));
       } else {
         v[i] = FromU62(si);
       }
     }
   }
 
  private:
-  constexpr static PrimeField FromU62(uint64_t input) {
+  template <
+      typename F2 = F,
+      std::enable_if_t<math::FiniteFieldTraits<F2>::kIsPrimeField>* = nullptr>
```

Comment:
Did you intend this function to be enabled for the case where F2 is packed prime field as well?